project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/MinimizationStageExecutorService.kt

```diff
@@ -40,12 +49,68 @@ class MinimizationStageExecutorService : MinimizationStageExecutor {
         context.withProgress {
             hierarchicalDD.minimize(it, hierarchy)
         }
-    }.onRight {
-        generalLogger.info { "End File level stage" }
-        statLogger.info { "File level stage result: success" }
+    }.logResult("File")
+
+    override suspend fun executeFunctionLevelStage(
+        context: IJDDContext,
+        functionLevelStage: FunctionLevelStage,
+    ) = either {
+        generalLogger.info { "Start Function level stage" }
+        statLogger.info {
+            "Function level stage settings. DDAlgorithm: ${functionLevelStage.ddAlgorithm::class.simpleName}"
+        }
+        val ddAlgorithm = functionLevelStage.ddAlgorithm.getDDAlgorithm()
+        val exceptionComparator = project
+            .service<MinimizationPluginSettings>()
+            .state
+            .exceptionComparingStrategy
+            .getExceptionComparator()
+        val lens = FunctionModificationLens()
+        val firstLevel = project
+            .service<MinimizationPsiManager>()
+            .findAllPsiWithBodyItems()
+        val propertyChecker = SameExceptionPropertyTester.create<PsiWithBodyDDItem>(
+            this@MinimizationStageExecutorService.project.service<BuildExceptionProviderService>(),
+            exceptionComparator,
+            lens,
+            context,
+        ).getOrElse {
+            generalLogger.error { "Property checker creation failed. Aborted" }
+            raise(MinimizationError.PropertyCheckerFailed)
+        }
+
+        firstLevel.logPsiElements()
+        context
+            .copy(currentLevel = firstLevel)
+            .withProgress {
+                ddAlgorithm.minimize(
+                    it,
+                    firstLevel,
+                    propertyChecker,
+                ).context
+            }
+    }.logResult("Function")
+
+    private suspend fun List<PsiWithBodyDDItem>.logPsiElements() {
+        if (!generalLogger.isDebugEnabled) {
+            return
+        }
+        val psiManagingService = project.service<MinimizationPsiManager>()
+        val text = mapNotNull {
+            psiManagingService.getPsiElementFromItem(it)?.let { readAction { it.text } }
+        }
+        generalLogger.debug {
```

Comment:
Here we can use trace, because messages are too big. But that's ok to ignore this comment